scipy/stats/_stats_py.py

```diff
@@ -10574,49 +10579,44 @@ def linregress(x, y, alternative='two-sided'):
     intercept (95%): 0.616950 +/- 0.544475
 
     """
-    TINY = 1.0e-20
-    x = np.asarray(x)
-    y = np.asarray(y)
+    np = array_namespace(x, y)  # needed for vecdot pre NumPy 2.0
 
-    if x.size == 0 or y.size == 0:
-        raise ValueError("Inputs must not be empty.")
-
-    if np.amax(x) == np.amin(x) and len(x) > 1:
-        raise ValueError("Cannot calculate a linear regression "
-                         "if all x values are identical")
+    TINY = 1.0e-20
 
-    n = len(x)
-    xmean = np.mean(x, None)
-    ymean = np.mean(y, None)
+    axis = -1
+    n = x.shape[axis]
+    xmean = np.mean(x, axis=axis, keepdims=True)
+    ymean = np.mean(y, axis=axis, keepdims=True)
 
     # Average sums of square differences from the mean
     #   ssxm = mean( (x-mean(x))^2 )
     #   ssxym = mean( (x-mean(x)) * (y-mean(y)) )
-    ssxm, ssxym, _, ssym = np.cov(x, y, bias=1).flat
+    x_ = _demean(x, xmean, axis=axis, xp=np)
+    y_ = _demean(y, ymean, axis=axis, xp=np)
+    xmean = np.squeeze(xmean, axis=-1)
+    ymean = np.squeeze(ymean, axis=-1)
+
+    ssxm = np.vecdot(x_, x_, axis=axis) / n
+    ssym = np.vecdot(y_, y_, axis=axis) / n
+    ssxym = np.vecdot(x_, y_, axis=axis) / n
 
     # R-value
     #   r = ssxym / sqrt( ssxm * ssym )
-    if ssxm == 0.0 or ssym == 0.0:
-        # If the denominator was going to be 0
-        r = np.asarray(np.nan if ssxym == 0 else 0.0)[()]
-    else:
-        r = ssxym / np.sqrt(ssxm * ssym)
-        # Test for numerical error propagation (make sure -1 < r < 1)
-        if r > 1.0:
-            r = 1.0
-        elif r < -1.0:
-            r = -1.0
+    degenerate = (ssxm == 0.0) | (ssym == 0.0)
+    r = xpx.apply_where(
+        ~degenerate,
+        (ssxym, ssxm, ssym),
+        lambda ssxym, ssxm, ssym: np.clip(ssxym / np.sqrt(ssxm * ssym), -1.0, 1.0),
+        lambda ssxym, ssxm, ssym: np.where(ssxym==0, np.nan, 0.0)
+    )
 
     slope = ssxym / ssxm
     intercept = ymean - slope*xmean
     if n == 2:
```

Comment:
OK to branch on array size until we want to support`marray`. (Maybe lazy arrays will be unhappy, too? We'll see.)